File: internal/tools/export.go
package tools

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/aleksandrgradoboev-svg/gt-data-1c/internal/refusal"
)

// pageSize — размер порции при выгрузке. Совпадает с потолком инструментов: больше
// одного запроса база всё равно не отдаёт.
const pageSize = 1000

// maxExportRows — предохранитель. Выгрузка в миллионы строк почти всегда означает
// забытый отбор, а не намерение: лучше упереться в понятный отказ, чем полчаса писать
// файл, который никто не откроет.
const maxExportRows = 500_000

type ExportInput struct {
	Base       string         `json:"base" jsonschema:"Имя базы 1С из реестра. Обязательно; перечень — bases с action=list"`
	Query      string         `json:"query" jsonschema:"Текст запроса на языке 1С. Только ВЫБРАТЬ. Для устойчивой выгрузки добавьте УПОРЯДОЧИТЬ"`
	Parameters map[string]any `json:"parameters,omitempty" jsonschema:"Параметры запроса: ключ без амперсанда"`
	Format     string         `json:"format,omitempty" jsonschema:"Формат файла: csv (умолчание) или jsonl"`
	Path       string         `json:"path,omitempty" jsonschema:"Куда записать файл. Пусто — каталог выгрузок в профиле пользователя"`
	MaxRows    int            `json:"max_rows,omitempty" jsonschema:"Предохранитель: остановиться после стольких строк (по умолчанию 500000)"`
}

func ExportTool() *mcp.Tool {
	return &mcp.Tool{
		Name: "export",
		Description: "Выгрузить результат запроса целиком в файл (csv или jsonl), не пропуская его " +
			"через ответ инструмента. Нужен, когда строк больше тысячи: обычный query отдаёт " +
			"страницу, а здесь сервер сам обходит все страницы и пишет файл, возвращая путь и " +
			"число строк. Ссылки в csv печатаются представлением, в jsonl — объектом с типом и " +
			"идентификатором. Для устойчивой выгрузки запрос должен содержать УПОРЯДОЧИТЬ.",
	}
}

func (s *Set) Export(ctx context.Context, _ *mcp.CallToolRequest, in ExportInput) (*mcp.CallToolResult, any, error) {
	if strings.TrimSpace(in.Query) == "" {
		return nil, nil, refusal.New(refusal.BadRequest, "текст запроса пуст", "поле query обязательно")
	}
	client, err := s.channelFor(in.Base)
	if err != nil {
		return nil, nil, err
	}

	format := strings.ToLower(strings.TrimSpace(in.Format))
	if format == "" {
		format = "csv"
	}
	if format != "csv" && format != "jsonl" {
		return nil, nil, refusal.New(refusal.BadRequest, "формат не распознан",
			"format="+in.Format, "допустимо: csv, jsonl")
	}

	limit := in.MaxRows
	if limit <= 0 {
		limit = maxExportRows
	}

	path, err := exportPath(in.Path, client.Base().Name, format)
	if err != nil {
		return nil, nil, err
	}
	file, err := os.Create(path)
	if err != nil {
		return nil, nil, refusal.New(refusal.Internal, "файл выгрузки не создан", err.Error())
	}
	defer file.Close()

	var (
		writer     *csv.Writer
		encoder    *json.Encoder
		columns    []string
		written    int
		offset     int
		totalKnown int
	)
	if format == "csv" {
		writer = csv.NewWriter(file)
		writer.Comma = ';' // Excel в русской локали ждёт точку с запятой
		defer writer.Flush()
	} else {
		encoder = json.NewEncoder(file)
	}

	started := time.Now()
	for {
		payload := map[string]any{"query": in.Query, "limit": pageSize, "offset": offset}
		if len(in.Parameters) > 0 {
			payload["parameters"] = in.Parameters
		}

		var reply queryReply
		if err := client.Tell(ctx, "query", payload, &reply); err != nil {
			os.Remove(path)
			return nil, nil, err
		}
		totalKnown = reply.ВсегоСтрок

		if written == 0 {
			columns = reply.Колонки
			if writer != nil {
				if err := writer.Write(columns); err != nil {
					return nil, nil, refusal.New(refusal.Internal, "заголовок не записан", err.Error())
				}
			}
		}

		for _, row := range reply.Строки {
			if err := writeRow(writer, encoder, columns, row); err != nil {
				return nil, nil, err
			}
			written++
			if written >= limit {
				break
			}
		}

		if !reply.ЕстьЕщё || written >= limit || len(reply.Строки) == 0 {
			offset = reply.СледующееСмещение
			break
		}
		offset = reply.СледующееСмещение
	}

	if writer != nil {
		writer.Flush()
		if err := writer.Error(); err != nil {
			return nil, nil, refusal.New(refusal.Internal, "файл выгрузки не дописан", err.Error())
		}
	}

	var b strings.Builder
	fmt.Fprintf(&b, "Выгружено строк: %d", written)
	if totalKnown > 0 {
		fmt.Fprintf(&b, " из %d", totalKnown)
	}
	fmt.Fprintf(&b, "\nФайл: %s\nФормат: %s, колонок %d, время %s",
		path, format, len(columns), time.Since(started).Round(time.Millisecond))
	if written >= limit && totalKnown > written {
		fmt.Fprintf(&b, "\n⚠ Выгрузка остановлена предохранителем на %d строках, в результате их %d. "+
			"Это не весь результат: уточните отбор или поднимите max_rows.", limit, totalKnown)
	}
	return text(b.String()), nil, nil
}

// writeRow пишет строку в выбранном формате.
//
// CSV разворачивает ссылку в представление: файл открывают глазами в таблице, и объект
// с типом и идентификатором там только мешает. JSONL сохраняет всё — его читает программа.
func writeRow(writer *csv.Writer, encoder *json.Encoder, columns []string, row map[string]any) error {
	if writer != nil {
		values := make([]string, 0, len(columns))
		for _, col := range columns {
			values = append(values, csvValue(row[col]))
		}
		if err := writer.Write(values); err != nil {
			return refusal.New(refusal.Internal, "строка не записана", err.Error())
		}
		return nil
	}
	if err := encoder.Encode(row); err != nil {
		return refusal.New(refusal.Internal, "строка не записана", err.Error())
	}
	return nil
}

func csvValue(value any) string {
	if ref, ok := value.(map[string]any); ok {
		представление, _ := ref["представление"].(string)
		return представление
	}
	if value == nil {
		return ""
	}
	return renderValue(value)
}

// exportPath выбирает имя файла: заданное вызывающим либо своё, с базой и временем.
func exportPath(explicit, base, format string) (string, error) {
	if strings.TrimSpace(explicit) != "" {
		if err := os.MkdirAll(filepath.Dir(explicit), 0o700); err != nil {
			return "", refusal.New(refusal.Internal, "каталог выгрузки не создан", err.Error())
		}
		return explicit, nil
	}

	dir, err := os.UserCacheDir()
	if err != nil || dir == "" {
		dir = "."
	}
	dir = filepath.Join(dir, "gt-data-1c", "exports")
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", refusal.New(refusal.Internal, "каталог выгрузок не создан", err.Error())
	}
	name := fmt.Sprintf("%s-%s.%s", base, time.Now().Format("20060102-150405"), format)
	return filepath.Join(dir, name), nil
}

File: crates/gf-data-1c/src/tools/export.rs
//! Выгрузка результата запроса в файл: когда строк больше, чем помещается в ответ.

use std::collections::BTreeMap;
use std::io::Write;
use std::path::PathBuf;

use serde::Deserialize;
use serde_json::{json, Map, Value};

use crate::refusal::{Kind, Refusal};

use super::data::QueryReply;
use super::Set;

/// Размер порции при выгрузке. Совпадает с потолком инструментов: больше одного запроса
/// база всё равно не отдаёт.
const PAGE_SIZE: i64 = 1000;

/// Предохранитель. Выгрузка в миллионы строк почти всегда означает забытый отбор, а не
/// намерение: лучше упереться в понятный отказ, чем полчаса писать файл, который никто
/// не откроет.
const MAX_EXPORT_ROWS: i64 = 500_000;

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct ExportInput {
    pub base: String,
    /// Текст запроса. Только ВЫБРАТЬ. Для устойчивой выгрузки добавьте УПОРЯДОЧИТЬ.
    pub query: String,
    pub parameters: BTreeMap<String, Value>,
    /// `csv` (умолчание) или `jsonl`.
    pub format: String,
    /// Куда положить файл. Пусто — каталог выгрузок в профиле пользователя.
    pub path: String,
    /// Потолок строк. Пусто — предохранитель на 500 000.
    pub max_rows: i64,
}

pub const NAME: &str = "export";

impl Set {
    /// Выгружает результат запроса в файл.
    pub fn export(&self, input: &ExportInput) -> Result<String, Refusal> {
        if input.query.trim().is_empty() {
            return Err(Refusal::new(
                Kind::BadRequest,
                "текст запроса пуст",
                "поле query обязательно",
            ));
        }
        let client = self.channel_for(&input.base)?;

        // Тот же гейт, что у query: выгрузка выполняет только текст построителя. Без
        // этого export — обход гейта в один ход: query рукописный текст отвергает, а
        // export с тем же текстом писал файл (найдено живой сессией 02.09.2026, модель
        // дыру нашла сама с четвёртой попытки).
        if !self.allow_raw_query && !self.gate.is_approved(&input.query) {
            return Err(Refusal::new(
                Kind::BadRequest,
                "текст запроса не собран построителем",
                "выгрузка выполняет только текст, который в этой сессии вернул query_build — \
                 дословно; написанный руками текст не выполняется, даже разобранный query_check",
            )
            .hint(
                "соберите запрос query_build (источник, поля, отбор, группировка, порядок) и \
                 выгружайте его текст как есть",
            ));
        }

        let format = input.format.trim().to_lowercase();
        let format = if format.is_empty() {
            "csv".to_string()
        } else {
            format
        };
        if format != "csv" && format != "jsonl" {
            return Err(Refusal::new(
                Kind::BadRequest,
                "формат не распознан",
                format!("format={}", input.format),
            )
            .hint("допустимо: csv, jsonl"));
        }

        let limit = if input.max_rows > 0 {
            input.max_rows
        } else {
            MAX_EXPORT_ROWS
        };

        let path = export_path(&input.path, &client.base().name, &format)?;
        let mut file = std::fs::File::create(&path)
            .map_err(|e| Refusal::new(Kind::Internal, "файл выгрузки не создан", e.to_string()))?;

        let started = std::time::Instant::now();
        let mut columns: Vec<String> = Vec::new();
        let mut written: i64 = 0;
        let mut offset: i64 = 0;
        // Присваивается в первой же итерации: цикл всегда делает хотя бы один запрос.
        let mut total_known: i64;

        loop {
            let mut payload = Map::new();
            payload.insert("query".into(), json!(input.query));
            payload.insert("limit".into(), json!(PAGE_SIZE));
            payload.insert("offset".into(), json!(offset));
            if !input.parameters.is_empty() {
                payload.insert("parameters".into(), json!(input.parameters));
            }

            let reply: QueryReply = match client.tell("query", &Value::Object(payload)) {
                Ok(r) => r,
                Err(e) => {
                    // Недописанный файл хуже отсутствующего: он выглядит результатом.
                    let _ = std::fs::remove_file(&path);
                    return Err(e);
                }
            };
            total_known = reply.rows_total;

            if written == 0 {
                columns = reply.columns.clone();
                if format == "csv" {
                    writeln!(file, "{}", csv_line(&columns)).map_err(|e| {
                        Refusal::new(Kind::Internal, "заголовок не записан", e.to_string())
                    })?;
                }
            }

            for row in &reply.rows {
                write_row(&mut file, &format, &columns, row)?;
                written += 1;
                if written >= limit {
                    break;
                }
            }

            if !reply.has_more || written >= limit || reply.rows.is_empty() {
                break;
            }
            offset = reply.next_offset;
        }

        file.flush()
            .map_err(|e| Refusal::new(Kind::Internal, "файл выгрузки не дописан", e.to_string()))?;

        let mut out = format!("Выгружено строк: {written}");
        if total_known > 0 {
            out.push_str(&format!(" из {total_known}"));
        }
        out.push_str(&format!(
            "\nФайл: {}\nФормат: {}, колонок {}, время {}ms",
            path.display(),
            format,
            columns.len(),
            started.elapsed().as_millis()
        ));
        if written >= limit && total_known > written {
            out.push_str(&format!(
                "\n⚠ Выгрузка остановлена предохранителем на {limit} строках, в результате их \
                 {total_known}. Это не весь результат: уточните отбор или поднимите max_rows."
            ));
        }
        Ok(out)
    }
}

/// Пишет строку в выбранном формате.
///
/// CSV разворачивает ссылку в представление: файл открывают глазами в таблице, и объект
/// с типом и идентификатором там только мешает. JSONL сохраняет всё — его читает программа.
fn write_row(
    file: &mut std::fs::File,
    format: &str,
    columns: &[String],
    row: &Map<String, Value>,
) -> Result<(), Refusal> {
    let line = if format == "csv" {
        let values: Vec<String> = columns
            .iter()
            .map(|c| csv_value(row.get(c).unwrap_or(&Value::Null)))
            .collect();
        csv_line(&values)
    } else {
        serde_json::to_string(row)
            .map_err(|e| Refusal::new(Kind::Internal, "строка не записана", e.to_string()))?
    };
    writeln!(file, "{line}")
        .map_err(|e| Refusal::new(Kind::Internal, "строка не записана", e.to_string()))
}

/// Собирает строку CSV. Разделитель — точка с запятой: Excel в русской локали ждёт
/// именно её, иначе весь файл ложится в одну колонку.
fn csv_line(values: &[String]) -> String {
    values
        .iter()
        .map(|v| csv_escape(v))
        .collect::<Vec<_>>()
        .join(";")
}

/// Экранирует значение по правилам CSV: кавычки удваиваются, поле берётся в кавычки,
/// если содержит разделитель, кавычку или перевод строки.
fn csv_escape(value: &str) -> String {
    if value.contains([';', '"', '\n', '\r']) {
        format!("\"{}\"", value.replace('"', "\"\""))
    } else {
        value.to_string()
    }
}

fn csv_value(value: &Value) -> String {
    if let Value::Object(m) = value {
        return m
            .get("представление")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_string();
    }
    if value.is_null() {
        return String::new();
    }
    super::data::render_value(value)
}

/// Выбирает имя файла: заданное вызывающим либо своё, с базой и временем.
fn export_path(explicit: &str, base: &str, format: &str) -> Result<PathBuf, Refusal> {
    if !explicit.trim().is_empty() {
        let p = PathBuf::from(explicit);
        if let Some(dir) = p.parent() {
            if !dir.as_os_str().is_empty() {
                std::fs::create_dir_all(dir).map_err(|e| {
                    Refusal::new(Kind::Internal, "каталог выгрузки не создан", e.to_string())
                })?;
            }
        }
        return Ok(p);
    }

    // Каталог прежний — см. registry::default_path.
    let dir = crate::journal::default_path()
        .parent()
        .map(|p| p.join("exports"))
        .unwrap_or_else(|| PathBuf::from("exports"));
    std::fs::create_dir_all(&dir)
        .map_err(|e| Refusal::new(Kind::Internal, "каталог выгрузок не создан", e.to_string()))?;

    let stamp = chrono::Local::now().format("%Y%m%d-%H%M%S");
    Ok(dir.join(format!("{base}-{stamp}.{format}")))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_set(tag: &str) -> Set {
        let dir = std::env::temp_dir().join(format!("gfdata-export-{}-{tag}", std::process::id()));
        let _ = std::fs::create_dir_all(&dir);
        let path = dir.join("bases.json");
        let _ = std::fs::remove_file(&path);
        let mut s = Set::new("0.1.0");
        s.registry_path = Some(path);
        s
    }

    fn с_базой(tag: &str) -> Set {
        let s = make_set(tag);
        let mut reg = s.registry().unwrap();
        reg.add(crate::registry::Base {
            name: "ut11".into(),
            url: "http://127.0.0.1:9/hs/gt-data".into(),
            ..Default::default()
        })
        .unwrap();
        s
    }

    #[test]
    fn выгрузка_рукописного_текста_это_обход_гейта() {
        let s = с_базой("гейт");
        let err = s
            .export(&ExportInput {
                base: "ut11".into(),
                query: "ВЫБРАТЬ Ссылка ИЗ Справочник.Номенклатура".into(),
                ..Default::default()
            })
            .unwrap_err();
        assert!(
            err.to_string().contains("не собран построителем"),
            "иначе export — дыра в гейте в один ход: {err}"
        );
    }

    #[test]
    fn незнакомый_формат_отвергается_с_перечнем() {
        let s = с_базой("формат");
        s.gate.approve("ВЫБРАТЬ 1");
        let err = s
            .export(&ExportInput {
                base: "ut11".into(),
                query: "ВЫБРАТЬ 1".into(),
                format: "xlsx".into(),
                ..Default::default()
            })
            .unwrap_err();
        assert!(err.to_string().contains("допустимо: csv, jsonl"), "{err}");
    }

    #[test]
    fn разделитель_csv_точка_с_запятой() {
        let line = csv_line(&["Код".to_string(), "Наименование".to_string()]);
        assert_eq!(line, "Код;Наименование", "Excel в русской локали ждёт «;»");
    }

    #[test]
    fn значение_с_разделителем_берётся_в_кавычки() {
        assert_eq!(csv_escape("Гвозди; шурупы"), "\"Гвозди; шурупы\"");
        assert_eq!(csv_escape("Труба \"Д\""), "\"Труба \"\"Д\"\"\"");
        assert_eq!(csv_escape("простое"), "простое");
    }

    #[test]
    fn ссылка_в_csv_сворачивается_до_представления() {
        let v = json!({
            "представление": "Гвозди",
            "тип": "CatalogRef.Номенклатура",
            "идентификатор": "abc"
        });
        assert_eq!(
            csv_value(&v),
            "Гвозди",
            "файл открывают глазами: тип и идентификатор там мешают"
        );
    }

    #[test]
    fn пустая_ячейка_в_csv_пуста_а_не_прочерк() {
        assert_eq!(csv_value(&Value::Null), "");
    }

    #[test]
    fn имя_файла_по_умолчанию_несёт_базу_и_время() {
        let p = export_path("", "ut11", "csv").unwrap();
        let name = p.file_name().unwrap().to_string_lossy().to_string();
        assert!(name.starts_with("ut11-"), "{name}");
        assert!(name.ends_with(".csv"), "{name}");
    }
}
